### PR TITLE
[SLP]Improve findReusedOrderedScalars and graph rotation.

### PR DESCRIPTION
**Pull request by alexey-bataev**
Patch syncs the code in findReusedOrderedScalars with cost
estimation/codegen. It tries to use similar logic to better determine
best order.
Before, it just tried to find previously vectorized node without
checking if it is possible to use the vectorized value in the shuffle.
Now it relies on the more generalized version. If it determines, that
a single vector must be reordered (using same mechanism, as codegen and
cost estimation), it generates better order.

The comparison between new/ref ordering:

Metric: SLP.NumVectorInstructions

Program                                                                                                                                                SLP.NumVectorInstructions
                                                                                                                                                       results                   results0 diff
                                                                                               test-suite :: MultiSource/Benchmarks/nbench/nbench.test   139.00                    140.00   0.7%
                                                                             test-suite :: MultiSource/Benchmarks/DOE-ProxyApps-C++/miniFE/miniFE.test   344.00                    346.00   0.6%
                                                                                        test-suite :: MultiSource/Benchmarks/FreeBench/pifft/pifft.test  1293.00                   1292.00  -0.1%
                                                                                test-suite :: External/SPEC/CFP2017rate/511.povray_r/511.povray_r.test  5176.00                   5170.00  -0.1%
                                                                                        test-suite :: External/SPEC/CFP2006/453.povray/453.povray.test  5173.00                   5167.00  -0.1%
                                                                                test-suite :: External/SPEC/CFP2017rate/510.parest_r/510.parest_r.test 11692.00                  11660.00  -0.3%
                                                                                     test-suite :: External/SPEC/CINT2006/464.h264ref/464.h264ref.test  1621.00                   1615.00  -0.4%
                                                                                             test-suite :: External/SPEC/CINT2006/403.gcc/403.gcc.test   795.00                    792.00  -0.4%
                                                                              test-suite :: External/SPEC/CFP2017rate/526.blender_r/526.blender_r.test 26499.00                  26338.00  -0.6%
                                                                                               test-suite :: MultiSource/Benchmarks/Bullet/bullet.test  7343.00                   7281.00  -0.8%
                                                                                          test-suite :: MultiSource/Applications/JM/ldecod/ldecod.test  1104.00                   1094.00  -0.9%
                                                                                          test-suite :: MultiSource/Applications/JM/lencod/lencod.test  2216.00                   2180.00  -1.6%
                                                                                            test-suite :: External/SPEC/CFP2006/433.milc/433.milc.test   787.00                    637.00 -19.1%

Less 0% is better.
Most of the benchmarks see more vectorized code. The first ones just
have shuffles removed.

The ordering analysis still may require some improvements (e.g. for
alternate nodes), but this one should be produce better results.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 1]
```diff
@@ -2418,7 +2418,8 @@ class BoUpSLP {
   std::optional<TargetTransformInfo::ShuffleKind>
   isGatherShuffledSingleRegisterEntry(
       const TreeEntry *TE, ArrayRef<Value *> VL, MutableArrayRef<int> Mask,
-      SmallVectorImpl<const TreeEntry *> &Entries, unsigned Part);
+      SmallVectorImpl<const TreeEntry *> &Entries, unsigned Part,
+      bool ForOrder);
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:2422]: Add ForOrder to the doxygen description

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 2]
```diff
@@ -2432,7 +2433,7 @@ class BoUpSLP {
   isGatherShuffledEntry(
       const TreeEntry *TE, ArrayRef<Value *> VL, SmallVectorImpl<int> &Mask,
       SmallVectorImpl<SmallVector<const TreeEntry *>> &Entries,
-      unsigned NumParts);
+      unsigned NumParts, bool ForOrder = false);
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:2436]: Add ForOrder to the doxygen description

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 3]
```diff
@@ -3779,65 +3780,169 @@ static void reorderOrder(SmallVectorImpl<unsigned> &Order, ArrayRef<int> Mask,
...
+    return CurrentOrder;
+  }
+  auto IsBroadcastMask = [](ArrayRef<int> Mask) {
+    int SingleElt = PoisonMaskElem;
+    return all_of(Mask, [&](int I) {
+      if (SingleElt == PoisonMaskElem && I != PoisonMaskElem)
+        SingleElt = I;
+      return I == PoisonMaskElem || I == SingleElt;
+    });
+  };
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3827]: Don't we have this anywhere else that we can reuse?

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 4]
```diff
@@ -3779,65 +3780,169 @@ static void reorderOrder(SmallVectorImpl<unsigned> &Order, ArrayRef<int> Mask,
...
+  if ((ExtractShuffles.empty() && IsBroadcastMask(Mask) &&
+       (Entries.size() != 1 ||
+        Entries.front().front()->ReorderIndices.empty())) ||
+      (GatherShuffles.empty() && IsBroadcastMask(ExtractMask)))
+    return std::nullopt;
+  SmallBitVector ShuffledSubMasks(NumParts);
+  auto TransformMaskToOrder = [&](MutableArrayRef<unsigned> CurrentOrder,
+                                  ArrayRef<int> Mask, int PartSz, int NumParts,
+                                  function_ref<unsigned(unsigned)> GetVF) {
+    for (int I : seq<int>(0, NumParts)) {
```
Comment (RKSimon) [inline llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3838]: why use seq instead of just a basic for loop?

**Review by RKSimon — COMMENTED** [review]
Please can you add better comments explaining the process

**Comment by alexey-bataev** [comment]
> Please can you add better comments explaining the process

Improves the rendering info, taken from gathered/buildvector nodes.
Currently, this info is taken only from vector nodes. E.g. if we have 2 nodes: vectorized <a1, a2, a3, a4> and gather/buildvector <a2, a1, a3, a4>, the reoder info is build only based on the vectorized node (how we need to reshuffle vectorized node to build the gathered/buildvector node). It does not take into account the neither the ordering within basic block, nor ordering because of other gathered/buildvector nodes, as codegen/cost model currently does.
The patch improves this process by using same approach, as codegen/cost model. Unfortunately, it cannot be unified with codegen/cost model approach directly (see processBuildVector template function, used for codegen/cost), because it requires some extra filtering, e.g. if the order comes from multiple nodes, it is not profitable to use it in many cases, since it may break the ordering that comes from the single node permutation. But the base process is still the same and relies on same functions, as codegen/cost model does.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp [hunk 5]
```diff
@@ -858,7 +858,7 @@ static void addMask(SmallVectorImpl<int> &Mask, ArrayRef<int> SubMask,
 /// values 3 and 7 respectively:
 /// before:  6 9 5 4 9 2 1 0
 /// after:   6 3 5 4 7 2 1 0
-static void fixupOrderingIndices(SmallVectorImpl<unsigned> &Order) {
+static void fixupOrderingIndices(MutableArrayRef<unsigned> Order) {
```
[the remark this answers is not in the document]
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:861]: Done.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 1 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:2422]: Done.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 2 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:2436]: Done

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 3 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3827]: Currently no. We have isSplat for values, but not for mask. For mask there is only ShuffleVectorInst::isZeroEltSplatMask, but here I need more common check, not only zero-based.

File: llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp (hunk 4 above)
Comment (alexey-bataev) [reply llvm/lib/Transforms/Vectorize/SLPVectorizer.cpp:3838]: https://llvm.org/docs/CodingStandards.html#use-range-based-for-loops-wherever-possible plus it simplifies the comparison of signed/unsigned data.

**Comment by alexey-bataev** [comment]
Ping!

**Review by RKSimon — APPROVED** [review]
LGTM

**Comment by eaeltsin** [comment]
Heads-up - this causes a miscompile in our tests. 

We are still working on the reduced reproducer, but so far the only difference in the miscompiled function looks like:
```
 ; Function Attrs: mustprogress nounwind uwtable
 define void @foo(ptr dead_on_unwind noalias nonnull writable sret(%"class.foo") align 8 %0, ptr nocapture noundef nonnull readonly align 8 dereferenceable(40) %1) local_unnamed_addr #0 align 32 {
   tail call void @bar(ptr noundef nonnull align 8 dereferenceable(40) %0, ptr noundef null) #9
   %3 = getelementptr inbounds i8, ptr %1, i64 24
   %4 = getelementptr inbounds i8, ptr %0, i64 24
   %5 = getelementptr inbounds i8, ptr %0, i64 16
   %6 = load i32, ptr %5, align 8
   %7 = load <4 x float>, ptr %3, align 8
   %8 = shufflevector <4 x float> %7, <4 x float> poison, <4 x i32> <i32 2, i32 3, i32 0, i32 1>
   %9 = fcmp olt <4 x float> %7, %8
   %10 = shufflevector <4 x float> %7, <4 x float> poison, <4 x i32> <i32 2, i32 3, i32 2, i32 3>
   %11 = shufflevector <4 x float> %7, <4 x float> poison, <4 x i32> <i32 0, i32 1, i32 0, i32 1>
   %12 = select <4 x i1> %9, <4 x float> %10, <4 x float> %11
-  %13 = shufflevector <4 x float> %12, <4 x float> poison, <4 x i32> <i32 2, i32 0, i32 3, i32 1>
+  %13 = shufflevector <4 x float> %12, <4 x float> poison, <4 x i32> <i32 1, i32 2, i32 3, i32 0>
   store <4 x float> %13, ptr %4, align 8
   %14 = or i32 %6, 15
   store i32 %14, ptr %5, align 8
   ret void
 }
```

**Comment by alexey-bataev** [comment]
> Heads-up - this causes a miscompile in our tests.
> 
> We are still working on the reduced reproducer, but so far the only difference in the miscompiled function looks like:
> 
> ```
>  ; Function Attrs: mustprogress nounwind uwtable
>  define void @foo(ptr dead_on_unwind noalias nonnull writable sret(%"class.foo") align 8 %0, ptr nocapture noundef nonnull readonly align 8 dereferenceable(40) %1) local_unnamed_addr #0 align 32 {
>    tail call void @bar(ptr noundef nonnull align 8 dereferenceable(40) %0, ptr noundef null) #9
>    %3 = getelementptr inbounds i8, ptr %1, i64 24
>    %4 = getelementptr inbounds i8, ptr %0, i64 24
>    %5 = getelementptr inbounds i8, ptr %0, i64 16
>    %6 = load i32, ptr %5, align 8
>    %7 = load <4 x float>, ptr %3, align 8
>    %8 = shufflevector <4 x float> %7, <4 x float> poison, <4 x i32> <i32 2, i32 3, i32 0, i32 1>
>    %9 = fcmp olt <4 x float> %7, %8
>    %10 = shufflevector <4 x float> %7, <4 x float> poison, <4 x i32> <i32 2, i32 3, i32 2, i32 3>
>    %11 = shufflevector <4 x float> %7, <4 x float> poison, <4 x i32> <i32 0, i32 1, i32 0, i32 1>
>    %12 = select <4 x i1> %9, <4 x float> %10, <4 x float> %11
> -  %13 = shufflevector <4 x float> %12, <4 x float> poison, <4 x i32> <i32 2, i32 0, i32 3, i32 1>
> +  %13 = shufflevector <4 x float> %12, <4 x float> poison, <4 x i32> <i32 1, i32 2, i32 3, i32 0>
>    store <4 x float> %13, ptr %4, align 8
>    %14 = or i32 %6, 15
>    store i32 %14, ptr %5, align 8
>    ret void
>  }
> ```

Waiting for the reproducer

**Comment by eaeltsin** [comment]
[reduced3.zip](https://github.com/llvm/llvm-project/files/14450582/reduced3.zip)

please see a zipped reproducer attached. please let me know if it is valid - maybe I overreduced it...

the difference I see after compiling with `-O3` is
```
   %5 = select <4 x i1> %2, <4 x float> %3, <4 x float> %4
-  %6 = shufflevector <4 x float> %5, <4 x float> poison, <4 x i32> <i32 2, i32 0, i32 3, i32 1>
+  %6 = shufflevector <4 x float> %5, <4 x float> poison, <4 x i32> <i32 1, i32 2, i32 3, i32 0>
   store <4 x float> %6, ptr %tmp.i.i4, align 8
```

**Comment by eaeltsin** [comment]
[reduced3.zip](https://github.com/llvm/llvm-project/files/14450664/reduced3.zip)

please see the reproducer attached. 

the difference I see after compiling with `-O3` is:
```
   %5 = select <4 x i1> %2, <4 x float> %3, <4 x float> %4
-  %6 = shufflevector <4 x float> %5, <4 x float> poison, <4 x i32> <i32 2, i32 0, i32 3, i32 1>
+  %6 = shufflevector <4 x float> %5, <4 x float> poison, <4 x i32> <i32 1, i32 2, i32 3, i32 0>
   store <4 x float> %6, ptr %tmp.i.i4, align 8
```

**Comment by alexey-bataev** [comment]
> [reduced3.zip](https://github.com/llvm/llvm-project/files/14450664/reduced3.zip)
> 
> please see the reproducer attached.
> 
> the difference I see after compiling with `-O3` is:
> 
> ```
>    %5 = select <4 x i1> %2, <4 x float> %3, <4 x float> %4
> -  %6 = shufflevector <4 x float> %5, <4 x float> poison, <4 x i32> <i32 2, i32 0, i32 3, i32 1>
> +  %6 = shufflevector <4 x float> %5, <4 x float> poison, <4 x i32> <i32 1, i32 2, i32 3, i32 0>
>    store <4 x float> %6, ptr %tmp.i.i4, align 8
> ```

Must be fixed by 45d82f33af9334907dde39cc69921c679299114e

**Pull request merged**